File: src/main/java/com/owncloud/android/ui/trashbin/TrashbinPresenter.java
/*
 * Nextcloud Android client application
 *
 * @author Tobias Kaminsky
 * Copyright (C) 2018 Tobias Kaminsky
 * Copyright (C) 2018 Nextcloud GmbH.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package com.owncloud.android.ui.trashbin;

import com.owncloud.android.R;
import com.owncloud.android.lib.resources.trashbin.model.TrashbinFile;

import java.io.File;
import java.util.List;

import static com.owncloud.android.datamodel.OCFile.ROOT_PATH;

/**
 * Coordinates between model and view: querying model, updating view, react to UI input
 */
public class TrashbinPresenter implements TrashbinContract.Presenter {

    private TrashbinContract.View trashbinView;
    private TrashbinRepository trashbinRepository;
    private String currentPath = ROOT_PATH;

    public TrashbinPresenter(TrashbinRepository trashbinRepository, TrashbinContract.View trashbinView) {
        this.trashbinRepository = trashbinRepository;
        this.trashbinView = trashbinView;
    }

    @Override
    public void enterFolder(String folder) {
        currentPath = folder;
        loadFolder();
    }

    @Override
    public boolean isRoot() {
        return !ROOT_PATH.equals(currentPath);
    }

    @Override
    public void navigateUp() {
        if (ROOT_PATH.equals(currentPath)) {
            trashbinView.close();
        } else {
            currentPath = new File(currentPath).getParent();

            loadFolder();
        }

        trashbinView.setDrawerIndicatorEnabled(ROOT_PATH.equals(currentPath));
    }

    @Override
    public void loadFolder() {
        trashbinRepository.getFolder(currentPath, new TrashbinRepository.LoadFolderCallback() {
            @Override
            public void onSuccess(List<Object> files) {
                trashbinView.showTrashbinFolder(files);
            }

            @Override
            public void onError(int error) {
                trashbinView.showError(error);
            }
        });
    }

    @Override
    public void restoreTrashbinFile(TrashbinFile file) {
        trashbinRepository.restoreFile(file, success -> {
            if (success) {
                trashbinView.removeFile(file);
            } else {
                trashbinView.showSnackbarError(R.string.trashbin_file_not_restored, file);
            }
        });
    }

    @Override
    public void removeTrashbinFile(TrashbinFile file) {
        trashbinRepository.removeTrashbinFile(file, success -> {
            if (success) {
                trashbinView.removeFile(file);
            } else {
                trashbinView.showSnackbarError(R.string.trashbin_file_not_deleted, file);
            }
        });
    }

    @Override
    public void emptyTrashbin() {
        trashbinRepository.emptyTrashbin(success -> {
            if (success) {
                trashbinView.removeAllFiles();
            } else {
                trashbinView.showError(R.string.trashbin_not_emptied);
            }
        });
    }
}

File: app/src/main/java/com/owncloud/android/ui/trashbin/TrashbinPresenter.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2024 TSI-mc <surinder.kumar@t-systems.com>
 * SPDX-FileCopyrightText: 2018 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2018 Nextcloud GmbH
 * SPDX-License-Identifier: AGPL-3.0-or-later OR GPL-2.0-only
 */
package com.owncloud.android.ui.trashbin

import com.owncloud.android.R
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.resources.trashbin.model.TrashbinFile
import com.owncloud.android.ui.trashbin.TrashbinContract.Presenter
import com.owncloud.android.ui.trashbin.TrashbinRepository.LoadFolderCallback
import java.io.File

/**
 * Coordinates between model and view: querying model, updating view, react to UI input
 */
class TrashbinPresenter(
    private val trashbinRepository: TrashbinRepository,
    private val trashbinView: TrashbinContract.View
) : Presenter {

    private var currentPath: String? = OCFile.ROOT_PATH

    override fun enterFolder(folder: String?) {
        currentPath = folder
        loadFolder()
    }

    override val isRoot: Boolean
        get() = OCFile.ROOT_PATH == currentPath

    override fun navigateUp() {
        if (isRoot) {
            trashbinView.close()
        } else {
            currentPath?.let {
                currentPath = File(it).parent
                loadFolder()
            }
        }
    }

    override fun loadFolder(onCompleted: () -> Unit, onError: () -> Unit) {
        trashbinRepository.getFolder(
            currentPath,
            object : LoadFolderCallback {
                override fun onSuccess(files: List<TrashbinFile?>?) {
                    trashbinView.showTrashbinFolder(files)
                    onCompleted()
                }

                override fun onError(error: Int) {
                    trashbinView.showError(error)
                    onError()
                }
            }
        )
        trashbinView.atRoot(isRoot)
    }

    override fun restoreTrashbinFile(files: Collection<TrashbinFile?>) {
        for (file in files) {
            trashbinRepository.restoreFile(
                file,
                object : TrashbinRepository.OperationCallback {
                    override fun onResult(success: Boolean) {
                        if (success) {
                            trashbinView.removeFile(file)
                        } else {
                            trashbinView.showSnackbarError(R.string.trashbin_file_not_restored, file)
                        }
                    }
                }
            )
        }
    }

    override fun removeTrashbinFile(files: Collection<TrashbinFile?>) {
        for (file in files) {
            trashbinRepository.removeTrashbinFile(
                file,
                object : TrashbinRepository.OperationCallback {
                    override fun onResult(success: Boolean) {
                        if (success) {
                            trashbinView.removeFile(file)
                        } else {
                            trashbinView.showSnackbarError(R.string.trashbin_file_not_deleted, file)
                        }
                    }
                }
            )
        }
    }

    override fun emptyTrashbin() {
        trashbinRepository.emptyTrashbin(object : TrashbinRepository.OperationCallback {
            override fun onResult(success: Boolean) {
                if (success) {
                    trashbinView.removeAllFiles()
                } else {
                    trashbinView.showError(R.string.trashbin_not_emptied)
                }
            }
        })
    }
}
